### PR TITLE
[CPF-692] Modelled Scheduling

### PR DESCRIPTION
**Pull request by KernelDeimos**
This PR adds the interface `foam.datetime.Schedule`, along with some implementations and the ability to use them in the existing Cron scheduler. Implementations added include `CronSchedule`, `TimeOfDaySchedule`, `IntervalSchedule`, and `OrSchedule`.

Here's an example journal entry for a job executing at 20 minute intervals and at 1PM every day:
```json
{
  "class": "foam.nanos.cron.Cron",
  "id":"hello-cron",
  "enabled": true,
  "schedule": {
    "class": "foam.datetime.OrSchedule",
    "delegates": [
      {
        "class": "foam.datetime.TimeOfDaySchedule",
        "time": {
          "class": "foam.datetime.TimeHMS",
          "hour": 13,
          "minute": 00,
          "second": 00
        }
      },
      {
        "class": "foam.datetime.IntervalSchedule",
        "duration": {
          "class": "foam.datetime.TimeHMS",
          "hour": 0,
          "minute": 20,
          "second": 00
        },
        "start": "2019-01-01T00:00:00.0Z"
      }
    ]
  },
  "description":"Says hello",
  "code":
  """
    System.out.println("Hello, Cron!");
  """
}
```

### CODE REVIEW

**Comment by jlhughes** [comment]
Is zero ignored/wildcard for Interval TimeHMS? 
The example above would suggest that, but Interval would suggest it's setting 0 to zero hour of day. 
`next.add(Calendar.HOUR_OF_DAY, getDuration().getHour());
`

**Comment by KernelDeimos** [comment]
IntervalSchedule holds a duration, so setting hour to zero means "zero hours added between each run"

**Review by jlhughes — COMMENTED** [review]
Looks great. 
Can you create some Test cases please.

File: src/foam/datetime/CronSchedule.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/datetime/CronSchedule.js]: 2019

File: src/foam/datetime/CronSchedule.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    next.set(Calendar.MINUTE, 0);
+    next.set(Calendar.SECOND, 0);
+    continue;
+  }
+  if ( getMinute() >= 0 && next.get(Calendar.MINUTE) != getMinute() ) {
+    next.add(Calendar.MINUTE, 1);
+    next.set(Calendar.SECOND, 0);
+    continue;
+  }
+  if( getSecond() >= 0 && next.get(Calendar.SECOND) != getSecond()) {
```
Comment (kgrgreer) [inline src/foam/datetime/CronSchedule.js]: Please see the foam style guide for spacing.

File: src/foam/datetime/IntervalSchedule.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+foam.CLASS({
```
Comment (kgrgreer) [inline src/foam/datetime/IntervalSchedule.js]: add copyright

File: src/foam/datetime/OrSchedule.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+foam.CLASS({
+  package: 'foam.datetime',
```
Comment (kgrgreer) [inline src/foam/datetime/OrSchedule.js]: I'm not sure this is important enough to be a top-level foam package. Maybe foam.util.schedule or just put in the foam.nanos.cron package.

File: src/foam/datetime/Schedule.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/datetime/Schedule.js]: 2019

File: src/foam/datetime/OrSchedule.js (hunk 4 above)
Comment (jlhughes) [reply src/foam/datetime/OrSchedule.js]: I prefer foam.nanos.cron over util.

**Comment by KernelDeimos** [comment]
Added test for IntervalSchedule - more tests pending. A little confused about the CLA thing caused by merging master.

**Comment by KernelDeimos** [comment]
CronSchedule copies the existing logic in Cron.js, so likely doesn't need a test. TimeIntervalSchedule and TimeOfDaySchedule both have tests now. OrSchedule is a simple "find minimum" function and has been tested manually.

**Comment by KernelDeimos** [comment]
@googlebot I consent.

File: src/foam/nanos/cron/CronSchedule.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.cron',
+  name: 'CronSchedule',
+  implements: [
+    'foam.nanos.cron.Schedule'
+  ],
+
+  javaImports: [
```
Comment (kgrgreer) [inline src/foam/nanos/cron/CronSchedule.js:14]: sort imports

File: src/foam/nanos/cron/CronSchedule.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    'foam.dao.DAO',
+    'foam.nanos.notification.Notification',
+    'foam.util.SafetyUtil',
+    'java.util.Calendar',
+    'java.util.concurrent.atomic.AtomicBoolean',
+    'java.util.Date'
+  ],
+
+  documentation: 'FOAM class that models a Cron script',
```
Comment (kgrgreer) [inline src/foam/nanos/cron/CronSchedule.js:23]: This documentation looks wrong. Isn't this just a schedule, not a cron script?

File: src/foam/nanos/cron/CronSchedule.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+  tableColumns: [
```
Comment (kgrgreer) [inline src/foam/nanos/cron/CronSchedule.js:25]: Do these columns exist?

File: src/foam/nanos/cron/CronSchedule.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    'id', 'enabled', 'server', 'description', 'lastDuration', 'status', 'run'
+  ],
+
+  searchColumns: ['id', 'description'],
```
Comment (kgrgreer) [inline src/foam/nanos/cron/CronSchedule.js:29]: Do these columns exist?

File: src/foam/nanos/cron/IntervalSchedule.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+foam.CLASS({
+  package: 'foam.nanos.cron',
+  name: 'IntervalSchedule',
+  implements: [
+    'foam.nanos.cron.Schedule'
+  ],
+
+  javaImports: [
+    'java.util.Date',
```
Comment (kgrgreer) [inline src/foam/nanos/cron/IntervalSchedule.js:15]: sort imports

File: src/foam/nanos/cron/Cron.js [hunk 11]
```diff
@@ -28,6 +28,11 @@ foam.CLASS({
   searchColumns: ['id', 'description'],
 
   properties: [
+    {
+      class: 'FObjectProperty',
+      of: 'foam.nanos.cron.Schedule',
+      name: 'schedule'
```
Comment (kgrgreer) [inline src/foam/nanos/cron/Cron.js:34]: Could you set the default value to a NeverSchedule singleton, then move the null pointer check below?

File: src/foam/nanos/cron/Schedule.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.INTERFACE({
+  package: 'foam.nanos.cron',
+  name: 'Schedule',
+  abstract: true,
```
Comment (kgrgreer) [inline src/foam/nanos/cron/Schedule.js:10]: Is abstract: true needed or used?

File: src/foam/nanos/cron/test/IntervalScheduleTest.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.cron.test',
+  name: 'IntervalScheduleTest',
+  extends: 'foam.nanos.test.Test',
+
```
Comment (kgrgreer) [inline src/foam/nanos/cron/test/IntervalScheduleTest.js]: Add documentation:.

File: src/foam/nanos/http/LiveScriptBundler.java [hunk 14]
```diff
@@ -0,0 +1,396 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/nanos/http/LiveScriptBundler.java:3]: 2019

File: src/foam/nanos/http/LiveScriptBundler.java [hunk 15]
```diff
@@ -0,0 +1,396 @@
+/**
```
Comment (kgrgreer) [inline src/foam/nanos/http/LiveScriptBundler.java:1]: This should be in a separate PR.

File: src/foam/nanos/http/LiveScriptBundler.java [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+import java.io.*;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.*;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import static java.nio.file.FileVisitResult.CONTINUE;
+import static java.nio.file.StandardWatchEventKinds.*;
+
+public class LiveScriptBundler implements WebAgent, ContextAware
```
Comment (kgrgreer) [inline src/foam/nanos/http/LiveScriptBundler.java:24]: You could extend ContextAwareSupport so you don't have to implement ContextAware directly.

File: src/foam/nanos/http/LiveScriptBundler.java [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+      }
+
+      name += ".js";
+      paths.add(name);
+    }
+
+    return paths;
+  }
+
+  private void log_(String evt, String msg) {
```
Comment (kgrgreer) [inline src/foam/nanos/http/LiveScriptBundler.java:356]: Why do this and not just use nanos logger?

File: src/foam/nanos/cron/CronSchedule.js (hunk 7 above)
Comment (KernelDeimos) [reply src/foam/nanos/cron/CronSchedule.js:23]: That's true. This documentation was taken from Cron.js as CronSchedule.js is just a carbon-copy of the schedule logic in Cron.js.

I didn't add documentation to the other schedules, so I'll add proper documentation for all of them now.

File: src/foam/nanos/cron/Cron.js (hunk 11 above)
Comment (KernelDeimos) [reply src/foam/nanos/cron/Cron.js:34]: The NeverSchedule singleton should disable a job, right? Right now the default behaviour (when schedule_==null) is to use the old Cron properties - this is my solution for making old DAO entries compatible.

Now that you have me thinking about it, I should be adapting the old properties into an instance of CronSchedule instead of keeping the now-duplicate logic in Cron.js.

File: src/foam/nanos/http/LiveScriptBundler.java (hunk 15 above)
Comment (KernelDeimos) [reply src/foam/nanos/http/LiveScriptBundler.java:1]: Oops, this is - not sure how it ended up here

**Comment by KernelDeimos** [comment]
I have no idea how LiveScriptBundler commits ended up in this branch. I looked at the tree and I just can't make sense of this.

I'm closing this PR, rebasing my branch, and opening a new one.

File: src/foam/nanos/http/LiveScriptBundler.java (hunk 17 above)
Comment (KernelDeimos) [reply src/foam/nanos/http/LiveScriptBundler.java:356]: This was updated before LiveScriptBundler was merged. Not sure what happened to this branch but it's a disaster now so I created a new one.
